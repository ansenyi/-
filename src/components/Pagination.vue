<template>
  <div class="pagination">
    <label for="page" class="select-label">
      <select id="page" class="select-size" ref="selectSize" @change="selectSize">
        <option value="5">每页10条</option>
        <option value="20">每页20条</option>
        <option value="30">每页30条</option>
      </select>
      <i id="first" class="select-down icon-down"></i>
    </label>
    <ul class="page-list">
      <li class="prev" v-show="curPage !== 1" @click="prevPage">
        <span><i class="icon-prev"></i></span>
      </li>
      <li class="first" v-show="totalPages > 1" @click="goPage(1)">
        <span :class="{'active': curPage === 1}">1</span>
      </li>
      <li class="ellipsis" v-show="prevClipped"><span>...</span></li>
      <li v-for="index in pages" :key="index" @click="goPage(index)">
        <span :class="{'active': curPage === index}">{{ index }}</span>
      </li>
      <li class="ellipsis" v-show="nextClipped"><span>...</span></li>
      <li class="last" @click="goPage(totalPages)">
        <span :class="{'active': curPage === totalPages}">{{ totalPages }}</span>
      </li>
      <li class="next" v-show="curPage !== totalPages" @click="nextPage">
        <span><i class="icon-next"></i></span>
      </li>
    </ul>
  </div>
</template>
<script>
  export default {
    name: 'pagination',
    props: {
      totalPages: {
        type: Number,
        default: 1,
        validator: i => i > 0
      },
      current: {
        type: Number,
        default: 1,
        validator: i => i > 0
      }
    },
    data () {
      return {
        nextClipped: false,
        prevClipped: false,
        curPage: this.current
      }
    },
    computed: {
      pages () {
        let ret = []
        this.curPage = this.current
        if (this.curPage > 3) {
          // 当前页码大于三时，显示当前页码的前2个
          ret.push(this.curPage - 2)
          ret.push(this.curPage - 1)
          this.prevClipped = this.curPage > 4
        } else {
          this.prevClipped = false
          for (let i = 2; i < this.curPage; i++) {
            ret.push(i)
          }
        }
        if (this.curPage !== this.totalPages && this.curPage !== 1) {
          ret.push(this.curPage)
        }
        if (this.curPage < (this.totalPages - 2)) {
          // 显示当前页码的后2个
          ret.push(this.curPage + 1)
          ret.push(this.curPage + 2)
          // 当前页与最后一页差距3以上时显示省略号
          this.nextClipped = this.curPage < (this.totalPages - 3)
        } else {
          this.nextClipped = false
          for (let i = (this.curPage + 1); i < this.totalPages; i++) {
            ret.push(i)
          }
        }
        return ret
      }
    },
    watch: {
      curPage (newVal) {
        console.log('当前页面currentPage：' + newVal)
        this.$emit('toPage', newVal)
        // this.$forceUpdate()
      }
    },
    methods: {
      selectSize () {
        let select = this.$refs.selectSize
        let index = select.selectedIndex
        let size = select[index].value
        this.$emit('changeSize', size)
      },
      prevPage () {
        this.curPage--
      },
      nextPage () {
        this.curPage++
      },
      goPage (index) {
        if (index !== this.curPage) {
          this.curPage = index
        }
      }
    }
  }
</script>

<style lang="stylus" scoped>
  @import '../common/stylus/mixin.styl'
  .pagination
    display flex
    justify-content center
    align-items center
    margin-top 20px
    .select-label
      select-label()
      &:before
        top 0
        right 0
        width 94px
        height 26px
        border-radius 3px
      .select-size
        height 26px
        color #444
        border 1px solid #999
        border-radius 3px
      .select-down
        top 4px
        right 4px
        color #999
    .page-list
      display flex
      li
        list-style none
        color #666
        //padding 0 3px
        &:not(.ellipsis)
          span
            display block
            min-width 24px
            height 25px
            margin 0 4px
            line-height 25px
            text-align center
            border 1px solid #999
            border-radius 3px
            text-shadow 0 1px 0 #fff
            cursor pointer
            background #fff
            i
              line-height 25px
          .active, span:hover
            color #fff
            background #0077d5
            /*border 1px solid #0077d5*/
</style>
