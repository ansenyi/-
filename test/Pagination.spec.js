/* eslint-disable no-undef,no-unused-vars */
import 'babel-polyfill'
import jest from 'jest-mock'
import { mount } from 'vue-test-utils'
import Pagination from '@/components/Pagination.vue'

describe('pagination', () => {
  function initComponent () {
    const wrapper = mount(Pagination, {
      propsData: {
        totalPages: 100,
        current: 5
      }
    })
    return wrapper
  }

  describe('props', () => {
    describe('type and default', () => {
      const totalPages = mount(Pagination).vm.$options.props.totalPages
      const current = mount(Pagination).vm.$options.props.current
      it('type Number', () => {
        expect(totalPages.type).toBe(Number)
        expect(current.type).toBe(Number)
      })
      it('default value', () => {
        expect(totalPages.default).toBe(1)
        expect(current.default).toBe(1)
      })
    })
    describe('set props', () => {
      const wrapper = initComponent()
      it('props value', () => {
        expect(wrapper.vm.totalPages).toBe(100)
        expect(wrapper.vm.current).toBe(5)
      })
    })
  })

  describe('select size emit event', () => {
    it('select change', () => {
      const wrapper = initComponent()
      const stub = jest.fn()
      wrapper.setMethods({selectSize: stub})
      const el = wrapper.find('select').trigger('change')
      expect(stub).toBeCalled()
    })
  })

  describe('watcher curPage', () => {
    let cmp = initComponent()
    let spy = jest.spyOn(console, 'log')
    afterEach(() => {
      spy.mockClear()
    })
    // it('is not called if value is empty (trimmed)', () => {
    //   cmp.setProps({current: ''})
    //   // cmp.setData({curPage: ''})
    //   cmp.vm.$nextTick(() => {
    //     expect(spy).not.toBeCalled()
    //     // next()
    //   })
    // })
    // it('is not called if values are the same', () => {
    //   cmp.setData({curPage: 5})
    //   cmp.vm.$nextTick(() => {
    //     spy.mockClear()
    //     cmp.setData({curPage: 5})
    //     cmp.vm.$nextTick(() => {
    //       spy = jest.spyOn(console, 'log')
    //       expect(spy).not.toBeCalled()
    //       // next()
    //     })
    //   })
    // })
    it('is called with the new value in other cases', () => {
      cmp.setData({curPage: 4})
      cmp.vm.$nextTick(() => {
        expect(spy).toBeCalled()
        // next()
      })
    })
    it('emit toPage', () => {
      const stub = jest.fn()
      cmp.vm.$on('toPage', stub)
      cmp.setData({curPage: 4})
      cmp.vm.$nextTick(() => {
        expect(stub).toBeCalledWith(4)
        // next()
      })
    })
  })

  describe('Simple Cases', () => {
    // false
    it('success', () => {
      const wrapper = initComponent()
      wrapper.update()
      expect(wrapper.element).toMatchSnapshot()
      expect(wrapper.findAll('li')).toHaveLength(11)
      expect(wrapper.find('li:first-of-type').hasClass('prev')).toBeTruthy()
      expect(wrapper.find('li:first-of-type').hasStyle('display', 'none')).toBeFalsy()
      expect(wrapper.find('li:last-of-type').hasStyle('display', 'none')).toBeFalsy()
      expect(wrapper.find('li:nth-of-type(2) span').text()).toBe('1')
      expect(wrapper.find('li:nth-of-type(3) span').text()).toBe('...')
      expect(wrapper.find('li:nth-of-type(3)').hasStyle('display', 'none')).toBeFalsy()
      expect(wrapper.find('li:nth-of-type(4) span').text()).toBe('3')
      expect(wrapper.find('.active').text()).toBe('5')
      expect(wrapper.find('li:nth-of-type(7) span').text()).toBe('6')
      expect(wrapper.find('li:nth-of-type(9) span').text()).toBe('...')
      expect(wrapper.find('li:nth-of-type(10) span').text()).toBe('100')
    })
    it('simulate next and prev button event right', () => {
      const wrapper = initComponent()
      wrapper.setProps({current: 4})
      // wrapper.update()
      expect(wrapper.findAll('li')).toHaveLength(11)
      expect(wrapper.find('.active').text()).toBe('4')
      expect(wrapper.find('li:nth-of-type(4)').text()).toBe('2')
      expect(wrapper.find('li:nth-of-type(8)').text()).toBe('6')
      expect(wrapper.find('li:nth-of-type(9)').text()).toBe('...')
      wrapper.setProps({current: 5})
      expect(wrapper.find('.active').text()).toBe('5')
    })
    it('simulate click page element', () => {
      const wrapper = initComponent()
      wrapper.setProps({current: 3})
      expect(wrapper.vm.prevClipped).toBeFalsy()
    })

    it('set last page 10', () => {
      const cmp = mount(Pagination, {
        propsData: {
          totalPages: 10,
          current: 10
        }
      })
      expect(cmp.find('li:last-of-type').hasStyle('display', 'none')).toBeTruthy()
      // 8
      expect(cmp.find('li:nth-of-type(4)').text()).toBe('8')
      expect(cmp.find('li:nth-of-type(7)').text()).toBe('10')
    })

    it('set totalPages 5', () => {
      const cmp = mount(Pagination, {
        propsData: {
          totalPages: 5,
          current: 1
        }
      })
      expect(cmp.element).toMatchSnapshot()
      expect(cmp.vm.prevClipped).toBeFalsy()
      // true
      // expect(cmp.findAll('.ellipsis')[0].hasStyle('display', 'none')).toBeTruthy()
      // expect(cmp.findAll('.ellipsis')[1].hasStyle('display', 'none')).toBeFalsy()
      expect(cmp.vm.nextClipped).toBeTruthy()
      expect(cmp.find('li:last-of-type').hasStyle('display', 'none')).toBeFalsy()
      cmp.setProps({current: 2})
      cmp.update()
      expect(cmp.vm.prevClipped).toBeFalsy()
      expect(cmp.vm.nextClipped).toBeFalsy()
      cmp.setProps({current: 3})
      cmp.update()
      expect(cmp.vm.prevClipped).toBeFalsy()
      expect(cmp.vm.nextClipped).toBeFalsy()
      cmp.setProps({current: 4})
      cmp.update()
      expect(cmp.vm.prevClipped).toBeFalsy()
      expect(cmp.vm.nextClipped).toBeFalsy()
      cmp.setProps({current: 5})
      cmp.update()
      expect(cmp.vm.prevClipped).toBeTruthy()
      expect(cmp.vm.nextClipped).toBeFalsy()
      expect(cmp.find('li:first-of-type').hasStyle('display', 'none')).toBeFalsy()
    })

    it('set totalPages 4', () => {
      const cmp = mount(Pagination, {
        propsData: {
          totalPages: 4,
          current: 1
        }
      })
      expect(cmp.vm.prevClipped).toBeFalsy()
      expect(cmp.vm.nextClipped).toBeFalsy()
    })
  })
})
