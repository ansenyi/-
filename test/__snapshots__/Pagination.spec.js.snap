// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pagination Simple Cases set totalPages 5 1`] = `
<div
  class="pagination"
>
  <label
    class="select-label"
    for="page"
  >
    <select
      class="select-size"
      id="page"
    >
      <option
        value="10"
      >
        每页10条
      </option>
       
      <option
        value="20"
      >
        每页20条
      </option>
       
      <option
        value="30"
      >
        每页30条
      </option>
    </select>
     
    <i
      class="select-down icon-down"
      id="first"
    />
  </label>
   
  <ul
    class="page-list"
  >
    <li
      class="prev"
      style="display: none;"
    >
      <span>
        <i
          class="icon-prev"
        />
      </span>
    </li>
     
    <li
      class="first"
    >
      <span
        class="active"
      >
        1
      </span>
    </li>
     
    <li
      class="ellipsis"
      style="display: none;"
    >
      <span>
        ...
      </span>
    </li>
     
    <li>
      <span
        class=""
      >
        2
      </span>
    </li>
    <li>
      <span
        class=""
      >
        3
      </span>
    </li>
     
    <li
      class="ellipsis"
    >
      <span>
        ...
      </span>
    </li>
     
    <li
      class="last"
    >
      <span
        class=""
      >
        5
      </span>
    </li>
     
    <li
      class="next"
    >
      <span>
        <i
          class="icon-next"
        />
      </span>
    </li>
  </ul>
</div>
`;

exports[`pagination Simple Cases success 1`] = `
<div
  class="pagination"
>
  <label
    class="select-label"
    for="page"
  >
    <select
      class="select-size"
      id="page"
    >
      <option
        value="10"
      >
        每页10条
      </option>
       
      <option
        value="20"
      >
        每页20条
      </option>
       
      <option
        value="30"
      >
        每页30条
      </option>
    </select>
     
    <i
      class="select-down icon-down"
      id="first"
    />
  </label>
   
  <ul
    class="page-list"
  >
    <li
      class="prev"
    >
      <span>
        <i
          class="icon-prev"
        />
      </span>
    </li>
     
    <li
      class="first"
    >
      <span
        class=""
      >
        1
      </span>
    </li>
     
    <li
      class="ellipsis"
      style=""
    >
      <span>
        ...
      </span>
    </li>
     
    <li>
      <span
        class=""
      >
        3
      </span>
    </li>
    <li>
      <span
        class=""
      >
        4
      </span>
    </li>
    <li>
      <span
        class="active"
      >
        5
      </span>
    </li>
    <li>
      <span
        class=""
      >
        6
      </span>
    </li>
    <li>
      <span
        class=""
      >
        7
      </span>
    </li>
     
    <li
      class="ellipsis"
    >
      <span>
        ...
      </span>
    </li>
     
    <li
      class="last"
    >
      <span
        class=""
      >
        100
      </span>
    </li>
     
    <li
      class="next"
    >
      <span>
        <i
          class="icon-next"
        />
      </span>
    </li>
  </ul>
</div>
`;
